src/backend/gl/src/device.rs

```diff
@@ -788,6 +810,48 @@ impl d::Device<B> for Device {
             vertex_buffers[vb.binding as usize] = Some(*vb);
         }
 
+        let mut uniforms = Vec::new();
+        {
+            let gl = &self.share.context;
+            let mut count = 0;
+            let mut uniform_max_size = 0;
+
+            gl.GetProgramiv(program, gl::ACTIVE_UNIFORMS, &mut count);
+            gl.GetProgramiv(
+                program,
+                gl::ACTIVE_UNIFORM_MAX_LENGTH,
+                &mut uniform_max_size,
+            );
+            for uniform in 0..count {
+                let mut length = 0;
+                let mut size = 0;
+                let mut utype = 0;
+                let mut name = Vec::with_capacity(uniform_max_size as usize);
+                name.set_len(uniform_max_size as usize - 1);
```

Comment:
`set_len` is unsafe, and we are not meeting one of it's in variants:
> The elements at old_len..new_len must be initialized.

 - https://doc.rust-lang.org/std/vec/struct.Vec.html

`glGetActiveUniform` might not init the whole range. This could cause UB in `Vec`'s `Drop` impl.

Reply:
Iterating through all of the active uniforms, one of them will initialize the rest of the vec, so I don't think this is an issue.

Reply:
Untrue. If all the active uniforms have a length less than `uniform_max_size - 1` then none of them will initialize the whole range.

Reply:
That's what this is querying:
```rust
            gl.GetProgramiv(
                program,
                gl::ACTIVE_UNIFORM_MAX_LENGTH,
                &mut uniform_max_size,
            );
```

Reply:
Hmmm, forgot the "ACTIVE" parts, although `glGetProgramiv` isn't to be trusted, as drivers are buggy: https://stackoverflow.com/questions/12555165/incorrect-value-from-glgetprogramivprogram-gl-active-uniform-max-length-outpa

Let me modify what I said just slightly: If any of the active uniforms have a length less than uniform_max_size - 1 then one of them will not initialize the whole range.

I'm afraid you aren't reusing the vector between iterations, instead making a new one each time.

Reply:
Assuming that's an old bug, and would have been fixed in the seven years since it was found *crosses fingers*, I hope we don't have to rely on that. This code is a little squirrely since we're relying on glGetActiveUniform to read up until a null terminator but giving it a string of size `uniform_max_size`. I've tested this code with many different variants of uniforms and it seems to be alright for now.